### ISSUE TITLE
Fortran: Doxygen confuses PRIVATE/PUBLIC statements for TYPEs defined in MODULEs (Origin: bugzilla #706864)

### ISSUE THREAD

**Issue by doxygen** [doxygen/doxygen#5273]
status NEW severity *normal* in component *general* for *---*
Reported in version *1.8.4* on platform *Other*
Assigned to: Dimitri van Heesch

On 2013-08-27 08:09:35 +0000, Stefan wrote:

    Hi there! 
    
    I just wanted to report that Doxygen 1.8.4 confuses Fortran's PUBLIC/PRIVATE
    for TYPEs defined in MODULEs. This happens for types with components
    declared as private however type-bound-procedures remaining public.
    
    Here is an example code: 
    > MODULE my_mod
    >
    >     IMPLICIT NONE
    >
    >     PRIVATE
    >
    >     TYPE my_public_type
    >     PRIVATE 
    >         INTEGER :: with_private_component
    >     CONTAINS
    >         !> A public type-bound-procedure.
    >         !! By mistake Doxygen 1.8.4 assumes it as private.
    >         PROCEDURE :: returns_private_component
    >     END TYPE
    >
    >     PUBLIC :: my_public_type
    >
    > CONTAINS
    >
    >     !> Private module procedure
    >     INTEGER FUNCTION returns_private_component(self)
    >         CLASS(my_public_type), INTENT(IN) :: self
    >
    >         returns_private_component = self%with_private_component
    >
    >     END FUNCTION
    >
    > END MODULE my_mod
    
    Thanks in advance. Cheers, 
    Stefan 
    
    Btw.: Why is a Fortran module listed as data type in Doxygen?

On 2013-08-27 18:18:25 +0000, albert wrote:

    Isn't returns_private_component private as well due to the private statement
    in the type definition part?

On 2013-08-27 21:12:37 +0000, Stefan wrote:

    Indeed, this is a little tricky in Fortran. 
    
    Although, the module procedure returns_private_compnent() (line 20) is
    declared private the type bound procedure returns_private_component() (line
    13) is by default public. Public/private statements of components and type-
    bound-procedures are treated separately.
    This is defined in section 4.5.5 of the standard. In particular see
    specification R446.
    
    Cheers, 
    Stefan

On 2013-08-28 16:47:09 +0000, albert wrote:

    Thanks for the information, have to study it but looks like quite a change
    required in the code.
    For the record the information is from the 2008 standard.

On 2013-10-08 15:53:05 +0000, Stefan wrote:

    True, I referred to the F08 standard. Nevertheless, that concept was
    introduced in Fortran 2003. Unfortunately, the Fortran standard is super
    cryptic and can hardly be read by us humans ;-).
    
    For F03: http://www.j3-fortran.org/doc/year/04/04-007.pdf
    
    See section 4.5.3.6 for components 
    and in particular NOTE 4.44 which reads:
    > The accessibility of a type-bound procedure is not affected by a PRIVATE
    statement in the component-part the accessibility of a data component is not
    affected by a PRIVATE statement in the type-bound-procedure-part.
    
    To summarize:
    TYPE ...
    <default access spec. for all components of that type>
        INTEGER, <access spec.> ...
    CONTAINS
    <default access spec. for all procedures of that type>
        PROCEDURE, <access spec.> ...
    END 
    
    Cheers, 
    Stefan

**Comment by PaulXiCao** [comment]
I can replicate this error even with a simpler example.
See the following

```fortranmodule test_m
  implicit none

  type test_t
    private
    integer :: a                ! interpreted as public
    integer, public :: b        ! interpreted as public
    integer, private :: c       ! interpreted as private
  end type test_t

end module test_m
```

One can see, that the initial `private` statement is totally neglected..
